File: backend/src/routes/query.routes.ts
import { Router } from 'express';
import { PrismaClient } from '@prisma/client';
import { decrypt } from '../services/encryption';
import { executeReadOnlyQuery, getSchemaSummary } from '../services/mysqlExecutor';
import { convertNlToSql } from '../services/nlToSql';

const router = Router();
const prisma = new PrismaClient();

// Run Query
router.post('/run', async (req, res) => {
    console.log(`[Query Route] Received request for connectionId: ${req.body.connectionId}, prompt: "${req.body.prompt}"`);
    try {
        const { connectionId, prompt } = req.body;

        if (!connectionId || !prompt) {
            return res.status(400).json({ error: "Missing connectionId or prompt" });
        }

        // 1. Fetch credentials
        const connection = await prisma.connection.findUnique({ where: { id: connectionId } });
        if (!connection) {
            return res.status(404).json({ error: "Connection not found" });
        }

        const password = decrypt(connection.password);
        const connectionParams = {
            host: connection.host,
            port: connection.port,
            database: connection.database,
            user: connection.username,
            password: password
        };

        // 2. Fetch Schema context
        const schema = await getSchemaSummary(connectionParams);
        console.log(`Schema fetched for connection ${connectionId}. Length: ${schema.length} chars.`);

        if (!schema || schema.includes("Could not fetch schema") || schema.includes("The database is empty")) {
            console.warn("Empty or invalid schema for connection:", connectionId);
        }

        // 3. Convert NL to SQL
        const result = await convertNlToSql(schema, prompt);

        if (!result.sql) {
            return res.status(422).json({
                error: result.explanation || "Could not generate SQL for this prompt.",
                explanation: result.explanation
            });
        }

        // 4. Execute SQL
        const data = await executeReadOnlyQuery(connectionParams, result.sql);

        // 5. Calculate basic metrics
        const totalRows = data.rows.length;
        // Simple numeric column detection for 'Total Value' or similar
        // This is a naive heuristic
        let numericSum = 0;
        if (totalRows > 0) {
            const firstRow = (data.rows as any[])[0];
            const numberFields = Object.keys(firstRow).filter(k => typeof firstRow[k] === 'number');
            if (numberFields.length > 0) {
                // Sum the first found number field as a sample metric
                numericSum = (data.rows as any[]).reduce((sum, row) => sum + (row[numberFields[0]] || 0), 0);
            }
        }

        res.json({
            sql: result.sql,
            explanation: result.explanation,
            columns: data.columns,
            rows: data.rows,
            metrics: {
                totalRows,
                approxSum: numericSum
            }
        });

    } catch (error: any) {
        console.error("Query Handler Error:", error);
        res.status(500).json({ error: error.message });
    }
});

export default router;

File: backend_python/app/routes/query_routes.py
from fastapi import APIRouter, Depends, HTTPException, Body
from fastapi.encoders import jsonable_encoder
from sqlalchemy.orm import Session
from app.models import get_db, Connection, QueryHistory
from app.services.encryption import decrypt
from app.services.mysql_executor import MySQLService
from app.services.nl_to_sql import convert_nl_to_sql, generate_data_insights
import json

router = APIRouter()

@router.post("/run")
async def run_query(connectionId: str = Body(...), prompt: str = Body(...), db: Session = Depends(get_db)):
    print(f"[Query Route] Received request for connectionId: {connectionId}, prompt: \"{prompt}\"")
    try:
        # 1. Fetch credentials
        connection = db.query(Connection).filter(Connection.id == connectionId).first()
        if not connection:
            raise HTTPException(status_code=404, detail="Connection not found")

        password = decrypt(connection.password)
        connection_params = {
            "host": connection.host,
            "port": connection.port,
            "database": connection.database,
            "user": connection.username,
            "password": password
        }

        # 2. Fetch Schema context
        schema = MySQLService.get_schema_summary(connection_params)
        print(f"Schema fetched for connection {connectionId}. Length: {len(schema)} chars.")

        if not schema or "Could not fetch schema" in schema or "The database is empty" in schema:
            print(f"Empty or invalid schema for connection: {connectionId}")

        # 3. Convert NL to SQL
        result = await convert_nl_to_sql(schema, prompt)

        if not result.get("sql"):
            # Point 7: Provide a friendly, informative explanation instead of failing.
            return {
                "sql": None,
                "explanation": result.get("explanation", "I couldn't process this request. Could you try rephrasing?"),
                "columns": [],
                "rows": [],
                "metrics": {"totalRows": 0, "approxSum": 0}
            }

        # 4. Execute SQL
        data = MySQLService.execute_read_only_query(connection_params, result["sql"])

        # 5. Calculate basic metrics
        rows = data["rows"]
        total_rows = len(rows)
        numeric_sum = 0
        
        if total_rows > 0:
            first_row = rows[0]
            # Detect numeric fields
            number_fields = [k for k, v in first_row.items() if isinstance(v, (int, float))]
            if number_fields:
                field = number_fields[0]
                numeric_sum = sum(float(row.get(field) or 0) for row in rows)

        # 6. Generate Real-time Data Insights (Rule 3)
        rich_explanation = result.get("explanation", "")
        
        # If we have data, let's get deep insights
        if total_rows > 0 and result.get("sql"):
            try:
                print(f"[Query Route] Generating deep data insights for {total_rows} rows...")
                data_insights = await generate_data_insights(prompt, result["sql"], rows, schema)
                
                # Merge the initial explanation (which has Reason/Available Data) with the data insights
                # We extract the first part of the explanation if it has the headers
                explanation_parts = rich_explanation.split("📊 Data Visualization & Insights:")
                header_part = explanation_parts[0] if explanation_parts else rich_explanation
                
                rich_explanation = f"{header_part.strip()}\n\n{data_insights.strip()}"
            except Exception as e:
                print(f"[Query Route] Insights generation failed: {e}")

        # 7. Save to History
        try:
            history_record = QueryHistory(
                user_id=connection.user_id or "default_user",
                conn_id=connection.id,
                conn_name=connection.name,
                prompt=prompt,
                sql_query=result["sql"],
                explanation=rich_explanation,
                columns=json.dumps(jsonable_encoder(data["columns"])),
                rows_data=json.dumps(jsonable_encoder(rows[:100])), # Save top 100 rows safely
                metrics=json.dumps({"totalRows": total_rows, "approxSum": numeric_sum})
            )
            db.add(history_record)
            db.commit()
            print(f"[Query Route] Saved query to history for {connection.name}")
        except Exception as e:
            print(f"[Query Route] Failed to save history: {e}")
            db.rollback()

        return {
            "sql": result["sql"],
            "explanation": rich_explanation,
            "columns": data["columns"],
            "rows": rows,
            "metrics": {
                "totalRows": total_rows,
                "approxSum": numeric_sum
            }
        }

    except HTTPException as e:
        raise e
    except Exception as e:
        print(f"Query Handler Error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))
